src/gluonts/evaluation/backtest.py

```diff
@@ -38,9 +38,148 @@
+
+    # ensure that the window size is small enough to allow rolling
+    assert(_num_items_in_rolling > window_size)
+
+    _remainder = _num_items_in_rolling % window_size
+    _remainder_date = end_time - _freq*_remainder
+    _num_window_iterations = int((_num_items_in_rolling - _remainder)/window_size)
+    if _remainder:
+        _num_window_iterations = _num_window_iterations + 1
+
+    # TODO test this
+    def remove_remainder(data):
+        data = data.copy()
+        start = data['start']
+        target = data['target']
+        end_date = start + _freq*(len(target)-1)
+        
+        # delete all values after the remainder date to allow these to be evaluated
+        if end_date > _remainder_date:
+            length_to_remainder = len(pd.date_range(start=start, end=_remainder_date, freq=_freq))
+            data['target'] = target[:length_to_remainder]
+        return data
+
+    # removes target values appearing after end_time
+    def truncate_end(data):
+        data = data.copy()
+        data['rolling'] = (start_time, end_time)
+
+        #calc number datapoints needed from start of timeseries until end of rolling test range
+        timerange = pd.date_range(
+            start=data['start'], 
+            end=end_time, 
+            freq=_freq
+        )
+        max_items = len(timerange)
+
+        # keep values until end of test range
+        if max_items < len(data['target']):
+            data['target'] = data['target'][:max_items]
+        
+        return data
+    
+    # generator to create rolling datasets
+    def perform_roll(dataset):
+        for timeseries in dataset:
+            windows_todo = _num_window_iterations
+            window_start_date = end_time
+            while windows_todo:
+                _ts = timeseries.copy()
+                
+                # 1. calc length of target 
+                target_length = len(_ts['target'])
+
+                # 2. calc new length of target
+                window_start_date = window_start_date - _freq * window_size
+                length_to_window = len(pd.date_range(start=_ts['start'], end=window_start_date, freq=_freq))
+                
+                # 3. if length of target < new length of target set flag "not touched" 
+                if target_length < length_to_window:
+                    # this avoids duplicate evaluations in the evaluator
+                    _ts['not_touched'] = True
+                else:
+                    _ts['target'] = _ts['target'][:length_to_window]
+                
+                # 4. yield the truncated/rolled timeseries even if it is "not touched" 
+                windows_todo = windows_todo - 1
+                yield _ts
+
+    # creates a generator for a evaluation dataset of the same 
+    # number timeseries as the rolling set
+    def generate_test(dataset):
+        for timeseries in dataset:
+            ts = timeseries.copy()
+            for _ in range(_num_window_iterations):
```

Comment:
if we perform in total N rolls then this will output N copies of each timeseries. This is to ensure the same amount of timeseries of the dataset used for prediction and evaluation.

Reply:
> This is to ensure the same amount of timeseries of the dataset used for prediction and evaluation.

Why should we enforce the same amount of time-series?

Reply:
This method is a little bit work in progress as it depends on how the evaluator works/should be changed, I only quickly skimmed the evaluator as of now.  
By providing this restrictions we (should) simplify the logic of the evaluator as it can compare timeseries on a 1-by-1 basis, i.e. 1 forecast to 1 real-timeseries without needing to calculate which real-timeseries belongs to which forecast.

Reply:
discussed offline, this will be remade for next commit